### PR TITLE
Typedoc comments for daml-types

### PR DESCRIPTION
**Pull request by robin-da**
### Pull Request Checklist

This converts comments to in daml-types to typedoc comments and adds some more documentation.

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: language-support/ts/daml-types/index.ts [hunk 1]
```diff
@@ -5,13 +5,21 @@ import * as jtv from '@mojotech/json-type-validation';
 /**
  * Interface for companion objects of serializable types. Its main purpose is
  * to describe the JSON encoding of values of the serializable type.
+ *
+ * @typeparam T The contract template type.
```
Comment (cocreature) [inline language-support/ts/daml-types/index.ts]:
```suggestion
 * @typeparam T The template type.
```

File: language-support/ts/daml-types/index.ts [hunk 2]
```diff
@@ -5,13 +5,21 @@ import * as jtv from '@mojotech/json-type-validation';
...
  */
 export interface Serializable<T> {
-  // NOTE(MH): This must be a function to allow for mutually recursive decoders.
+  /**
+   * The decoder for a contract of template T.
+   *
+   * NB: This is a function to allow for mutually recursive decoders.
```
Comment (cocreature) [inline language-support/ts/daml-types/index.ts]: That seems like an implementation detail. Do we want to keep this in the docs? In fact should we hide the whole `Serializable` thing?

File: language-support/ts/daml-types/index.ts [hunk 3]
```diff
@@ -22,6 +30,11 @@ export const STATIC_IMPLEMENTS_SERIALIZABLE_CHECK = <T>(_: Serializable<T>) => {
 /**
  * Interface for objects representing DAML templates. It is similar to the
  * `Template` type class in DAML.
+ *
+ * @typeparam T The contract template type.
```
Comment (cocreature) [inline language-support/ts/daml-types/index.ts]:
```suggestion
 * @typeparam T The template type.
```

File: language-support/ts/daml-types/index.ts [hunk 4]
```diff
@@ -30,18 +43,43 @@ export interface Template<T extends object, K = unknown, I extends string = stri
 }
 
 /**
- * Interface for objects representing DAML choices. It is similar to the
- * `Choice` type class in DAML.
+ * Interface for objects representing DAML choices.
+ *
+ * It is similar to the `Choice` type class in DAML.
+ *
+ * @typeparam T The contract template type.
```
Comment (cocreature) [inline language-support/ts/daml-types/index.ts]:
```suggestion
 * @typeparam T The template type.
```

File: language-support/ts/daml-types/index.ts [hunk 5]
```diff
@@ -30,18 +43,43 @@ export interface Template<T extends object, K = unknown, I extends string = stri
 }
 
 /**
- * Interface for objects representing DAML choices. It is similar to the
- * `Choice` type class in DAML.
+ * Interface for objects representing DAML choices.
+ *
+ * It is similar to the `Choice` type class in DAML.
```
Comment (cocreature) [inline language-support/ts/daml-types/index.ts]: There is actually no `Choice` typeclass anymore (only a constraint synonym). And even if there is, we never advertised this in DAML and treated it more like an implementation detail so I would just remove this.

File: language-support/ts/daml-types/index.ts [hunk 6]
```diff
@@ -30,18 +43,43 @@ export interface Template<T extends object, K = unknown, I extends string = stri
...
+ *
+ * @typeparam T The contract template type.
+ * @typeparam K The contract key type.
+ * @typeparam C The choice type.
+ * @typeparam R The choice return type.
+ *
  */
 export interface Choice<T extends object, C, R, K = unknown> {
+  /**
+   * Returns the contract template to which this choice belongs.
```
Comment (cocreature) [inline language-support/ts/daml-types/index.ts]:
```suggestion
   * Returns the template to which this choice belongs.
```

File: language-support/ts/daml-types/index.ts [hunk 7]
```diff
@@ -30,18 +43,43 @@ export interface Template<T extends object, K = unknown, I extends string = stri
...
+   */
   template: () => Template<T, K>;
+  /**
+   * Returns a decoder to decode the choice arguments.
+   */
```
Comment (cocreature) [inline language-support/ts/daml-types/index.ts]: Should we make the decoders internal?

File: language-support/ts/daml-types/index.ts [hunk 8]
```diff
@@ -79,42 +120,54 @@ export const Unit: Serializable<Unit> = {
...
 }
 
 /**
- * The counterpart of DAML's `Numeric` type. We represent `Numeric`s as string
- * in order to avoid a loss of precision. The string must match the regular
- * expression `-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?`.
+ * The counterpart of DAML's `Numeric` type.
+ *
+ * We represent `Numeric`s as string in order to avoid a loss of precision. The string must match
+ * the regular expression `-?(?:0|[1-9]\d*)(?:\.\d+)?(?:[eE][+-]?\d+)?`.
```
Comment (cocreature) [inline language-support/ts/daml-types/index.ts]: Do we actually validate this on the client side anywhere or do you only get the error from the JSON API. If the latter, would it make sense to introduce validation somewhere to get more readable errors?

File: language-support/ts/daml-types/index.ts [hunk 9]
```diff
@@ -79,42 +120,54 @@ export const Unit: Serializable<Unit> = {
...
+ * Companion function of the [[Numeric]] type.
  */
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
 export const Numeric = (_: number): Serializable<Numeric> =>
   ({
     decoder: jtv.string,
   })
 
+/**
+ * Companion function of the [[Decimal]] type.
```
Comment (cocreature) [inline language-support/ts/daml-types/index.ts]: Is this really a function?

File: language-support/ts/daml-types/index.ts (hunk 2 above)
Comment (robin-da) [reply language-support/ts/daml-types/index.ts]: My thinking was that if someone is looking at the documentation of daml-types, then they probably want to build something on their own. So the serializable thing might be of interest to them. But I wouldn't mind either to make it internal.

File: language-support/ts/daml-types/index.ts (hunk 5 above)
Comment (robin-da) [reply language-support/ts/daml-types/index.ts]: ok, then let's remove this line.

File: language-support/ts/daml-types/index.ts (hunk 7 above)
Comment (robin-da) [reply language-support/ts/daml-types/index.ts]: see above.

File: language-support/ts/daml-types/index.ts (hunk 8 above)
Comment (robin-da) [reply language-support/ts/daml-types/index.ts]: I don't know, @hurryabit might know ?

File: language-support/ts/daml-types/index.ts (hunk 9 above)
Comment (robin-da) [reply language-support/ts/daml-types/index.ts]: a function without an argument ... changed it to object.

File: language-support/ts/daml-types/index.ts (hunk 8 above)
Comment (hurryabit) [reply language-support/ts/daml-types/index.ts]: Currently, we only validate data coming from the JSON API. We only validate that each `Numeric` is a string and nothing beyond that.

Long term, we should also validate the data we send to the JSON API to be able to produce better error messages if we were to send malformed data.

File: language-support/ts/daml-types/index.ts [hunk 10]
```diff
@@ -204,16 +268,22 @@ export const ContractId = <T>(_t: Serializable<T>): Serializable<ContractId<T>>
 
 /**
  * The counterpart of DAML's `Optional T` type.
+ *
+ * @typeparam T The type of the optionally present value.
```
Comment (hurryabit) [inline language-support/ts/daml-types/index.ts]: We should give a very brief explanation, that just using `null | T` would not work if you have nested `Optional`s. That's where the seemingly complicated `OptionalInner` type is coming from. See https://docs.daml.com/json-api/lf-value-specification.html#optional for an explanation of the `Optional` encoding.

Comment (shayne-fletcher) [reply language-support/ts/daml-types/index.ts]: See also the last note of the "Primitive DAML Types" section of https://docs.daml.com/daml2ts/index.html.

**Pull request merged**